### PR TITLE
[XLN] Fix XLN

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
[This commit](https://github.com/huggingface/transformers/commit/87e6e4fe5c7e65cb69e70306f22de6daf16b6e14) broke the XLNet docstyle explaining what the output of the `create_mask` function should be.

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Still a draft, will be fixing  #21626 via a deprecation cycle

File: src/transformers/models/xlnet/configuration_xlnet.py [hunk 1]
```diff
@@ -57,7 +57,9 @@ class XLNetConfig(PretrainedConfig):
         untie_r (`bool`, *optional*, defaults to `True`):
             Whether or not to untie relative position biases
         attn_type (`str`, *optional*, defaults to `"bi"`):
-            The attention type used by the model. Set `"bi"` for XLNet, `"uni"` for Transformer-XL.
+            The attention type used by the model. Set `"bi"` for XLNet.`"uni"` is not longer supperted. If you want to
+            use Transformer-XL, use a
+            [`Transformer-XLModel`](https://huggingface.co/docs/transformers/model_doc/transfo-xl).
```
Comment (sgugger) [inline src/transformers/models/xlnet/configuration_xlnet.py:62]: The link is very weird. I'd just do
```suggestion
            The attention type used by the model. Set `"bi"` for XLNet.`"uni"` is no longer supported, use [Transformer-XL](./transfo-xl) instead.
```

File: src/transformers/models/xlnet/modeling_tf_xlnet.py [hunk 2]
```diff
@@ -629,8 +629,10 @@ def call(
         # Attention mask
         # causal attention mask
         if self.attn_type == "uni":
-            attn_mask = self.create_mask(qlen, mlen)
-            attn_mask = attn_mask[:, :, None, None]
+            raise FutureWarning(
```
Comment (sgugger) [inline src/transformers/models/xlnet/modeling_tf_xlnet.py:632]:
```suggestion
            raise warnings.warn(
```

File: src/transformers/models/xlnet/modeling_tf_xlnet.py [hunk 3]
```diff
@@ -629,8 +629,10 @@ def call(
...
+                f"`attn_type` {self.attn_type} is not supported anymore and will be removed in version 5 of transformers."
+                "If you want to use transformers XL use the [Transformer XL model](https://huggingface.co/docs/transformers/model_doc/transfo-xl) "
+            )
```
Comment (sgugger) [inline src/transformers/models/xlnet/modeling_tf_xlnet.py:635]: Same comment as above + add FutureWarning here.

File: src/transformers/models/xlnet/modeling_xlnet.py [hunk 4]
```diff
@@ -1028,8 +1030,10 @@ def relative_positional_encoding(self, qlen, klen, bsz=None):
             # beg, end = klen - 1, -qlen
             beg, end = klen, -qlen
         elif self.attn_type == "uni":
-            # beg, end = klen - 1, -1
-            beg, end = klen, -1
+            raise FutureWarning(
```
Comment (sgugger) [inline src/transformers/models/xlnet/modeling_xlnet.py:1033]: Same comment as above.

File: src/transformers/models/xlnet/modeling_tf_xlnet.py [hunk 5]
```diff
@@ -629,8 +629,10 @@ def call(
         # Attention mask
         # causal attention mask
         if self.attn_type == "uni":
-            attn_mask = self.create_mask(qlen, mlen)
-            attn_mask = attn_mask[:, :, None, None]
```
Comment (sgugger) [inline src/transformers/models/xlnet/modeling_tf_xlnet.py:633]: Why is this removed?

**Review by sgugger — COMMENTED** [review]
Thanks for working on this. Not sure we should completely remove the old code since we are just deprecating this attention type. Also is it actually used in some checkpoints online? If that's the case, we can't just deprecate without fixing the checkpoints (to move to a Transformer XL model?)
